Userland/Applications/PixelPaint/ImageProcessor.cpp

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright (c) 2022, kleines Filmröllchen <filmroellchen@serenityos.org>.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "ImageProcessor.h"
+
+namespace PixelPaint {
+
+FilterApplicationCommand::FilterApplicationCommand(NonnullRefPtr<Filter> filter, NonnullRefPtr<Layer> target_layer)
+    : m_filter(move(filter))
+    , m_target_layer(move(target_layer))
+{
+}
+
+void FilterApplicationCommand::execute()
+{
+    m_filter->apply(m_target_layer->content_bitmap(), m_target_layer->content_bitmap());
+    m_filter->m_editor->gui_event_loop().deferred_invoke([strong_this = NonnullRefPtr(*this)]() {
+        // HACK: we can't tell strong_this to not be const
+        (*const_cast<NonnullRefPtr<Layer>*>(&strong_this->m_target_layer))->did_modify_bitmap(strong_this->m_target_layer->rect());
+        strong_this->m_filter->m_editor->did_complete_action();
+    });
+}
+
+static Singleton<ImageProcessor> s_image_processor;
+
+ImageProcessor::ImageProcessor()
+    : m_command_queue(MUST(Queue::try_create()))
+    , m_processor_thread(Threading::Thread::construct([this]() {
+        while (true) {
+            if (auto next_command = m_command_queue.try_dequeue(); !next_command.is_error()) {
+                next_command.value()->execute();
+            } else {
+                m_wakeup_mutex.lock();
+                m_wakeup_variable.wait_while([this]() { return m_command_queue.weak_used() == 0; });
+                m_wakeup_mutex.unlock();
```

Comment:
```suggestion
                MutexLocker locker (m_wakeup_mutex);
                m_wakeup_variable.wait_while([this]() { return m_command_queue.weak_used() == 0; });
```

Reply:
(Same for the locking after this too)

Reply:
Nope, because that unlocks too late. We want to keep the lock as small as possible as it's only used here for correct condvar protocol; the mutex doesn't actually protect anything.

Reply:
Then just
```c++
{
    MutexLocker locker );
    m_wakeup_variable.wait_while([this]() { return m_command_queue.weak_used() == 0; });
}
```
is also an option, or this:
```c++
m_wakeup_variable.wait_while(
    [this, locker = MutexLocker{m_wakeup_mutex}] {
        return m_command_queue.weak_used() == 0;
    }
);
```